Initial version of automation framework.

michael-berlin opened the pull request:
@alainjobart @aaijazi 

Let's agree on the basics of the automation framework. Please let me know if something should be changed.

Suggest order of files to review:
- automation.proto
- scheduler_test.go
- resharding_task_test.go
- scheduler.go
- resharding_task.go
- rest of the files

**Original commit message:**

The automation framework allows to automate cluster operations which
require a series of manual steps e.g. resharding.

A Cluster Operation has a list of task containers which are processed
sequentially. Each task container can contain one or more tasks which
will be executed in parallel.

Here's an example of a cluster operation with two task containers. The
second task container has two tasks:
- step 1
- step 2a | step 2b

If the task container contains one task, the task can emit new task
containers which will be inserted after the current task container. This
mechanism is used to fully expand Cluster Operations by special tasks
which emit new task containers e.g. "ReshardingTask".

This patchset implements the minimal steps to automate "resharding". It
lacks extensive testing of error corner cases which will be added in
later commits.

aaijazi commented on proto/automation.proto:
```diff
@@ -0,0 +1,91 @@
+
+// TaskContainer holds one or more task which may be executed in parallel.
+// "concurrency", if > 0, limits the amount of concurrently executed tasks.
+// TODO(mberlin): Find a better name. "TaskGroup"?
```
TaskContainer honestly seems fine to me.

michael-berlin replied:
DONE

I removed the TODO.

aaijazi commented on proto/automation.proto:
```diff
@@ -0,0 +1,91 @@
+  // Runtime data.
+  string id = 3;
+  TaskState state = 4;
+  // Set after state advanced to DONE. If empty, the task did succeed.
```
If output is empty, or error?

michael-berlin replied:
DONE

This was meant for "error", but then I added "output" later. I moved it to "error".

aaijazi commented on go/vt/automation/testing_emit_echo_task.go:
```diff
@@ -0,0 +1,19 @@
+// Copyright 2015, Google Inc. All rights reserved.
```
I don't see the value in this being its own file - combine all testing tasks into one file?

michael-berlin replied:
I'm a big fan of one class per file. Do we have any guidelines around this?

yaoshengzhe replied:
I think one class per file is good, it makes people easy to navigate your code even with limited IDE support. The only case I would prefer to put multiple classes in a single file is probably some message classes. Back to this case, it looks like this class is added for testing purpose, could we add "_test" prefix so that it does not count in go coverage and also it makes sure these test code won't in the production binaries.

michael-berlin replied:
I expressed the testing code nature of this class with the prefix "testing_".

The suffix "_test" may mislead people to think this is a test?

yaoshengzhe replied:
sorry I mean "_test" suffix, I believe it is Go's convention: Files with the suffix _test.go are only compiled and run by the go test tool. 

http://stackoverflow.com/questions/25161774/what-are-conventions-for-filenames-in-go

michael-berlin replied:
The problem with these Testing* tasks is that they are referenced in the factory function which instantiates tasks from the protobuf-description:

https://github.com/youtube/vitess/commit/20c10dbe7d19c462935a16002da33550668e2e1d#diff-3a3658336f82b97340ba072879b56fd9R166 

Therefore, we cannot compile them as test code only.

If so: When running tests, we would have to extend the factory method to be aware of the Testing* tasks as well. Can you think of an easy way to do this?

yaoshengzhe replied:
I think we use register pattern a lot. Basically, you could add a Register(name string, taskCreator func() Task) function, so that the test task could call the Register func in its init func. Therefore, the createTaskInstance func will look like getRegisteredTaskCreator(name)(), which you get a registered task creator and then create a task. This is my suggestion.

michael-berlin replied:
DONE

Shengzhe, thanks for the example! The ability to set functions so easily was the missing piece :-)

I've moved all code (including the two test tasks) into a new file "testutils_test.go".

aaijazi commented on go/vt/automation/scheduler.go:
```diff
@@ -0,0 +1,258 @@
+				break
+			}
+
+			log.Infof("Task: %v (%v/%v) running. Details: %v", taskProto.Name, clusterOp.Id, taskProto.Id, taskProto)
```
When should tasks be marked as running? Will each task be responsible for doing that itself?

michael-berlin replied:
No, that is missing here. I'll add it.

michael-berlin replied:
DONE

Added line for changing the task state to RUNNING.

aaijazi commented on go/vt/automation/scheduler.go:
```diff
@@ -0,0 +1,258 @@
+	for i := 0; i < len(clusterOp.SerialTasks); i++ {
+		taskContainer := clusterOp.SerialTasks[i]
+		for _, taskProto := range taskContainer.ParallelTasks {
+			if taskProto.State == pb.TaskState_DONE {
```
When would we be expecting state to be done already? Is this for retrying?

michael-berlin replied:
This is when resuming a checkpoint.

aaijazi commented on go/vt/automation/scheduler.go:
```diff
@@ -0,0 +1,258 @@
+	log.Infof("ClusterOperation: %v running. Details: %v", clusterOp.Id, clusterOp)
+
+	var lastTaskError string
+	for i := 0; i < len(clusterOp.SerialTasks); i++ {
```
Does range not work for this, because the slice is being extended?

michael-berlin replied:
Yes, range creates a copy of the slice and wouldn't see the appends.

aaijazi commented on go/vt/automation/scheduler.go:
```diff
@@ -0,0 +1,258 @@
+	s.state = stateRunning
+	s.mu.Unlock()
+
+	s.pendingOpsWg.Add(1)
```
This seems like a strange way to use a waitgroup - you're just waiting for a single operation to finish? I think it makes more sense to use an empty channel then: `done := make(chan struct{})`. And then, when you're finished: `defer close(done)`.

michael-berlin replied:
In the future, there may be more than one operation.

Given the "may" in this situation, I tried to change the code to your suggestion, it would make the overall code more complicated because of the following corner case:

Let's assume this execution:
s := NewScheduler()
s.ShutdownAndWait()

ShutdownAndWait() shouldn't block wait for the "done" channel to be closed now because the Scheduler is not in a "Running" or "ShuttingDown" state. But I would say these extra checks make the code more complicated.

Instead, I've moved the code pendingOpsWg.Add(1) and the launch of the go function into an extra method:

startProcessRequestsLoop()

Now it should be more clear that there may be more than one process.

aaijazi replied:
Oh, interesting, this is a good point. I think you should make it clear in the comments that there is a very good reason for doing what you're doing. Something like:

// Use a wait-group instead of just a done channel, because we want
// to be able to shut down the scheduler even if Run() was never executed.

michael-berlin replied:
DONE

aaijazi commented on go/vt/automation/task.go:
```diff
@@ -0,0 +1,16 @@
+// Tasks can emit new tasks by returning them. The new tasks will be added
+// to the execution graph directly after this task.
+type Task interface {
+	run(parameters map[string]string) (newTaskContainers []*pb.TaskContainer, output string, err error)
```
One thing that seems to be missing how we'd handle registering parameters. It would be useful/necessary to be able to expose the parameters that a Cluster Operation / Task depends on, in an eventual API.

michael-berlin replied:
I agree. I've added the following to this interface:

	// requiredParameters() returns a list of parameter keys which must be provided as input for run().
	requiredParameters() []string

The framework will take care of checking that all required parameters are specified. Tasks only have to return a list of required parameters through this interface method.

aaijazi replied:
Nice. The new requiredParameters feels a little clumsy, but I think it should work fine.

aaijazi commented on go/vt/automation/resharding_task_test.go:
```diff
@@ -0,0 +1,26 @@
+	reshardingTask := ReshardingTask{}
+
+	newTaskContainers, _, _ := reshardingTask.run(map[string]string{
+		"source_shard_rdonly_list": "rdonly1",
```
It doesn't matter for the purpose of this test, but I think for documentation purposes, you should make the rdonly list read more like real tablet aliases.

michael-berlin replied:
DONE

aaijazi commented on go/vt/automation/resharding_task.go:
```diff
@@ -0,0 +1,123 @@
+type ReshardingTask struct {
+}
+
+func splitShardListIntoKeyspaceAndShards(shardsWithKeyspace []string) (string, []string, error) {
```
Is there no existing function to do this?

michael-berlin replied:
DONE

Thanks for being persistent here! I've found topo.ParseKeyspaceShardString which parses on "keyspace/shard" string.

I've changed the code to use it.

aaijazi commented on go/vt/automation/resharding_task.go:
```diff
@@ -0,0 +1,123 @@
+)
+
+// ReshardingTask is a cluster operation which allows to increase the number of shards.
+type ReshardingTask struct {
```
Probably is more clear to call this HorizontalReshardingTask.

michael-berlin replied:
DONE

Renamed to "HorizontalReshardingTask".

aaijazi commented on go/vt/automation/resharding_task.go:
```diff
@@ -0,0 +1,123 @@
+	reparentTasks := NewTaskContainer()
+	for _, destShard := range destShards {
+		newMaster := selectAnyTabletFromShardByType(destShard, "master")
+		AddTask(reparentTasks, "ForceReparent", map[string]string{
```
There is no ForceReparent task right now - I imagine that's intentional? Doesn't that cause the cluster operation to fail?

michael-berlin replied:
DONE

Yes, that's right. I've commented the tasks and added TODOs. I'll add the implementation when I integrate it into your new worker.py test / create a copy of it.

aaijazi commented:
Overall, this seems reasonable to me. I'd like to see more comments, smaller functions, more unit tests... but those are probably not relevant for a prototype...

In the future though, it would be nice to break up PR so that you do multiple small PR instead of one large one. Thorough code review is hard at 1200 lines :)

michael-berlin commented:
@aaijazi 

I've addressed your comments and tried to push one commit for each change. Can you please review the new changes?

Unfortunately, the original commit got rewritten during some rebase action :-( Therefore, it's more difficult to go through your old comments :(

aaijazi commented:
LGTM!

michael-berlin commented:
I would like to merge this state for now and iterate over it later.

@aaijazi Can you please have a last look at the commits starting from "Remove vtctl and vtworker tasks." (Somehow, the history got lost again. Sorry about that :( )

I've made several changes to bump the test coverage.

aaijazi commented on go/vt/automation/scheduler_test.go:
```diff
@@ -0,0 +1,250 @@
+	if err != nil {
+		t.Fatalf("Failed to create scheduler: %v", err)
+	}
+	scheduler.setTaskCreator(testingTaskCreator)
```
It might makes sense to pull this code out into a function, otherwise I can see it being really easy to forget to set the task creator.

Instead: 

`scheduler := newTestScheduler()`

michael-berlin replied:
DONE

I also didn't like that it's a lot of boiler plate code.

I've added *testing.T as parameter to move error checking into the function as well.

This exposed several places where I accidentally used the wrong "err" variable or didn't check it after reusing it. Lesson learned: don't reuse "err", maybe even use scoping to make sure it doesn't get accidentally reused in checks.

aaijazi commented on go/vt/automation/scheduler_test.go:
```diff
@@ -0,0 +1,250 @@
+
+	waitForClusterOperation(t, scheduler, enqueueResponse.Id, expectedOutput, "")
+
+	scheduler.ShutdownAndWait()
```
Slightly cleaner to write this as a defer, up by scheduler creation:

```
scheduler := newTestScheduler()
defer scheduler.ShutdownAndWait()
```

michael-berlin replied:
DONE

aaijazi commented:
A couple of minor things, but otherwise LGTM
